### PR TITLE
sync: output full description for nested test cases

### PR DESCRIPTION
**Pull request by ErikSchierboom**
Some problems defined in the problem-specifications repo are defined using nesting. configlet handles these nicely, but the description of the `tests.toml` files only contains the description of the test case itself, which can be confusing if test cases in different nested objects have the same description.

This PR writes the full description to the `tests.toml` file. For example, `Test group description → Test case description` for a singly-nested test case.

Closes https://github.com/exercism/configlet/issues/202

While working on this PR, I found that there are two exercises that contain unneeded nesting: https://github.com/exercism/problem-specifications/pull/1798 I've deferred updating the tests until after that PR has been merged, to prevent unnecessary work.

### CODE REVIEW

File: src/sync/probspecs.nim
Comment (ErikSchierboom) [inline src/sync/probspecs.nim:13]: We need to keep track of the parents to iterate over their descriptions later on.

File: src/sync/probspecs.nim
Comment (ErikSchierboom) [inline src/sync/probspecs.nim:68]: If there is no matching, there is only one parent and it does not contain a description. I could also rewrite this to do, which should also work:

```suggestion
  if testCase.parents.len > 2:
```

The current version is a bit more fool-proof, should any nesting ever forget to include the `description`.

File: src/sync/probspecs.nim
Comment (ErikSchierboom) [inline src/sync/probspecs.nim:84]: I don't know if this is the nicest way to do this, or should I make the argument mutable?

**Review by ee7 — COMMENTED** [review]
### Indication of nesting

First, is there a better way for us to indicate the nesting?

With the current approach, some instances of `description` become extremely long, and it's difficult to see the underlying structure. Here's the `tests.toml` for `custom-set`, where we reach 194 characters in a line:

<details>
  <summary>Click to expand</summary>

```toml
# This is an auto-generated file. Regular comments will be removed when this
# file is regenerated. Regenerating will not touch any manually added keys,
# so comments can be added in a "comment" key.

[20c5f855-f83a-44a7-abdd-fe75c6cf022b]
description = "Returns true if the set contains no elements → sets with no elements are empty"

[d506485d-5706-40db-b7d8-5ceb5acf88d2]
description = "Returns true if the set contains no elements → sets with elements are not empty"

[759b9740-3417-44c3-8ca3-262b3c281043]
description = "Sets can report if they contain an element → nothing is contained in an empty set"

[f83cd2d1-2a85-41bc-b6be-80adbff4be49]
description = "Sets can report if they contain an element → when the element is in the set"

[93423fc0-44d0-4bc0-a2ac-376de8d7af34]
description = "Sets can report if they contain an element → when the element is not in the set"

[c392923a-637b-4495-b28e-34742cd6157a]
description = "A set is a subset if all of its elements are contained in the other set → empty set is a subset of another empty set"

[5635b113-be8c-4c6f-b9a9-23c485193917]
description = "A set is a subset if all of its elements are contained in the other set → empty set is a subset of non-empty set"

[832eda58-6d6e-44e2-92c2-be8cf0173cee]
description = "A set is a subset if all of its elements are contained in the other set → non-empty set is not a subset of empty set"

[c830c578-8f97-4036-b082-89feda876131]
description = "A set is a subset if all of its elements are contained in the other set → set is a subset of set with exact same elements"

[476a4a1c-0fd1-430f-aa65-5b70cbc810c5]
description = "A set is a subset if all of its elements are contained in the other set → set is a subset of larger set with same elements"

[d2498999-3e46-48e4-9660-1e20c3329d3d]
description = "A set is a subset if all of its elements are contained in the other set → set is not a subset of set that does not contain its elements"

[7d38155e-f472-4a7e-9ad8-5c1f8f95e4cc]
description = "Sets are disjoint if they share no elements → the empty set is disjoint with itself"

[7a2b3938-64b6-4b32-901a-fe16891998a6]
description = "Sets are disjoint if they share no elements → empty set is disjoint with non-empty set"

[589574a0-8b48-48ea-88b0-b652c5fe476f]
description = "Sets are disjoint if they share no elements → non-empty set is disjoint with empty set"

[febeaf4f-f180-4499-91fa-59165955a523]
description = "Sets are disjoint if they share no elements → sets are not disjoint if they share an element"

[0de20d2f-c952-468a-88c8-5e056740f020]
description = "Sets are disjoint if they share no elements → sets are disjoint if they share no elements"

[4bd24adb-45da-4320-9ff6-38c044e9dff8]
description = "Sets with the same elements are equal → empty sets are equal"

[f65c0a0e-6632-4b2d-b82c-b7c6da2ec224]
description = "Sets with the same elements are equal → empty set is not equal to non-empty set"

[81e53307-7683-4b1e-a30c-7e49155fe3ca]
description = "Sets with the same elements are equal → non-empty set is not equal to empty set"

[d57c5d7c-a7f3-48cc-a162-6b488c0fbbd0]
description = "Sets with the same elements are equal → sets with the same elements are equal"

[dd61bafc-6653-42cc-961a-ab071ee0ee85]
description = "Sets with the same elements are equal → sets with different elements are not equal"

[06059caf-9bf4-425e-aaff-88966cb3ea14]
description = "Sets with the same elements are equal → set is not equal to larger set with same elements"

[8a677c3c-a658-4d39-bb88-5b5b1a9659f4]
description = "Unique elements can be added to a set → add to empty set"

[0903dd45-904d-4cf2-bddd-0905e1a8d125]
description = "Unique elements can be added to a set → add to non-empty set"

[b0eb7bb7-5e5d-4733-b582-af771476cb99]
description = "Unique elements can be added to a set → adding an existing element does not change the set"

[893d5333-33b8-4151-a3d4-8f273358208a]
description = "Intersection returns a set of all shared elements → intersection of two empty sets is an empty set"

[d739940e-def2-41ab-a7bb-aaf60f7d782c]
description = "Intersection returns a set of all shared elements → intersection of an empty set and non-empty set is an empty set"

[3607d9d8-c895-4d6f-ac16-a14956e0a4b7]
description = "Intersection returns a set of all shared elements → intersection of a non-empty set and an empty set is an empty set"

[b5120abf-5b5e-41ab-aede-4de2ad85c34e]
description = "Intersection returns a set of all shared elements → intersection of two sets with no shared elements is an empty set"

[af21ca1b-fac9-499c-81c0-92a591653d49]
description = "Intersection returns a set of all shared elements → intersection of two sets with shared elements is a set of the shared elements"

[c5e6e2e4-50e9-4bc2-b89f-c518f015b57e]
description = "Difference (or Complement) of a set is a set of all elements that are only in the first set → difference of two empty sets is an empty set"

[2024cc92-5c26-44ed-aafd-e6ca27d6fcd2]
description = "Difference (or Complement) of a set is a set of all elements that are only in the first set → difference of empty set and non-empty set is an empty set"

[e79edee7-08aa-4c19-9382-f6820974b43e]
description = "Difference (or Complement) of a set is a set of all elements that are only in the first set → difference of a non-empty set and an empty set is the non-empty set"

[c5ac673e-d707-4db5-8d69-7082c3a5437e]
description = "Difference (or Complement) of a set is a set of all elements that are only in the first set → difference of two non-empty sets is a set of elements that are only in the first set"

[c45aed16-5494-455a-9033-5d4c93589dc6]
description = "Union returns a set of all elements in either set → union of empty sets is an empty set"

[9d258545-33c2-4fcb-a340-9f8aa69e7a41]
description = "Union returns a set of all elements in either set → union of an empty set and non-empty set is the non-empty set"

[3aade50c-80c7-4db8-853d-75bac5818b83]
description = "Union returns a set of all elements in either set → union of a non-empty set and empty set is the non-empty set"

[a00bb91f-c4b4-4844-8f77-c73e2e9df77c]
description = "Union returns a set of all elements in either set → union of non-empty sets contains all unique elements"
``` 

</details>

If we stick with this approach, we could consider using multi-line strings for long lines. I'd also like to have a bikeshedding session on the separator character itself - I'm don't love ` → `, but I don't know what's better.

But I think there's a better way to indicate the nesting. The options I can see:
1. The current approach
2. Add another special property to each test case, called e.g. `category`.
3. Use a TOML comment (`#`).
4. Use nested tables, and consider indenting each UUID table.

Here is option 3, which I think this is a step in the right direction:

<details>
  <summary>Click to expand</summary>

```toml
# This is an auto-generated file. Regular comments will be removed when this
# file is regenerated. Regenerating will not touch any manually added keys,
# so comments can be added in a "comment" key.

# Returns true if the set contains no elements
[20c5f855-f83a-44a7-abdd-fe75c6cf022b]
description = "sets with no elements are empty"

[d506485d-5706-40db-b7d8-5ceb5acf88d2]
description = "sets with elements are not empty"

# Sets can report if they contain an element
[759b9740-3417-44c3-8ca3-262b3c281043]
description = "nothing is contained in an empty set"

[f83cd2d1-2a85-41bc-b6be-80adbff4be49]
description = "when the element is in the set"

[93423fc0-44d0-4bc0-a2ac-376de8d7af34]
description = "when the element is not in the set"

# A set is a subset if all of its elements are contained in the other set
[c392923a-637b-4495-b28e-34742cd6157a]
description = "empty set is a subset of another empty set"

[5635b113-be8c-4c6f-b9a9-23c485193917]
description = "empty set is a subset of non-empty set"

[832eda58-6d6e-44e2-92c2-be8cf0173cee]
description = "non-empty set is not a subset of empty set"

[c830c578-8f97-4036-b082-89feda876131]
description = "set is a subset of set with exact same elements"

[476a4a1c-0fd1-430f-aa65-5b70cbc810c5]
description = "set is a subset of larger set with same elements"

[d2498999-3e46-48e4-9660-1e20c3329d3d]
description = "set is not a subset of set that does not contain its elements"

# Sets are disjoint if they share no elements
[7d38155e-f472-4a7e-9ad8-5c1f8f95e4cc]
description = "the empty set is disjoint with itself"

[7a2b3938-64b6-4b32-901a-fe16891998a6]
description = "empty set is disjoint with non-empty set"

[589574a0-8b48-48ea-88b0-b652c5fe476f]
description = "non-empty set is disjoint with empty set"

[febeaf4f-f180-4499-91fa-59165955a523]
description = "sets are not disjoint if they share an element"

[0de20d2f-c952-468a-88c8-5e056740f020]
description = "sets are disjoint if they share no elements"

# Sets with the same elements are equal
[4bd24adb-45da-4320-9ff6-38c044e9dff8]
description = "empty sets are equal"

[f65c0a0e-6632-4b2d-b82c-b7c6da2ec224]
description = "empty set is not equal to non-empty set"

[81e53307-7683-4b1e-a30c-7e49155fe3ca]
description = "non-empty set is not equal to empty set"

[d57c5d7c-a7f3-48cc-a162-6b488c0fbbd0]
description = "sets with the same elements are equal"

[dd61bafc-6653-42cc-961a-ab071ee0ee85]
description = "sets with different elements are not equal"

[06059caf-9bf4-425e-aaff-88966cb3ea14]
description = "set is not equal to larger set with same elements"

# Unique elements can be added to a set
[8a677c3c-a658-4d39-bb88-5b5b1a9659f4]
description = "add to empty set"

[0903dd45-904d-4cf2-bddd-0905e1a8d125]
description = "add to non-empty set"

[b0eb7bb7-5e5d-4733-b582-af771476cb99]
description = "adding an existing element does not change the set"

# Intersection returns a set of all shared elements
[893d5333-33b8-4151-a3d4-8f273358208a]
description = "intersection of two empty sets is an empty set"

[d739940e-def2-41ab-a7bb-aaf60f7d782c]
description = "intersection of an empty set and non-empty set is an empty set"

[3607d9d8-c895-4d6f-ac16-a14956e0a4b7]
description = "intersection of a non-empty set and an empty set is an empty set"

[b5120abf-5b5e-41ab-aede-4de2ad85c34e]
description = "intersection of two sets with no shared elements is an empty set"

[af21ca1b-fac9-499c-81c0-92a591653d49]
description = "intersection of two sets with shared elements is a set of the shared elements"

# Difference (or Complement) of a set is a set of all elements that are only in the first set
[c5e6e2e4-50e9-4bc2-b89f-c518f015b57e]
description = "difference of two empty sets is an empty set"

[2024cc92-5c26-44ed-aafd-e6ca27d6fcd2]
description = "difference of empty set and non-empty set is an empty set"

[e79edee7-08aa-4c19-9382-f6820974b43e]
description = "difference of a non-empty set and an empty set is the non-empty set"

[c5ac673e-d707-4db5-8d69-7082c3a5437e]
description = "difference of two non-empty sets is a set of elements that are only in the first set"

# Union returns a set of all elements in either set
[c45aed16-5494-455a-9033-5d4c93589dc6]
description = "union of empty sets is an empty set"

[9d258545-33c2-4fcb-a340-9f8aa69e7a41]
description = "union of an empty set and non-empty set is the non-empty set"

[3aade50c-80c7-4db8-853d-75bac5818b83]
description = "union of a non-empty set and empty set is the non-empty set"

[a00bb91f-c4b4-4844-8f77-c73e2e9df77c]
description = "union of non-empty sets contains all unique elements"
```

</details>

(We'd have to tweak the wording in the header about comments).

Any thoughts? Can we think of any other options?

### Unnecessary nesting of test cases

As you described for e.g. `isogram` and `prime-factors`, we have the problem where every case is prefixed needlessly:

```toml
[a0e97d2d-669e-47c7-8134-518a1e2c4555]
description = "Check if the given string is an isogram → empty string"

[9a001b50-f194-4143-bc29-2af5ec1ef652]
description = "Check if the given string is an isogram → isogram with only lower case characters"

[...]
```

```toml
[924fc966-a8f5-4288-82f2-6b9224819ccd]
description = "returns prime factors for the given input number → no factors"

[17e30670-b105-4305-af53-ddde182cb6ad]
description = "returns prime factors for the given input number → prime number"

[...]
```

However, I believe there are many more exercises with the same problem. Here is an incomplete list:
- [`alphametics`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/alphametics/canonical-data.json)
- [`book-store`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/book-store/canonical-data.json)
- [`etl`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/etl/canonical-data.json)
- [`gigasecond`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/gigasecond/canonical-data.json)
- [`nucleotide-count`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/nucleotide-count/canonical-data.json)
- [`ocr-numbers`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/ocr-numbers/canonical-data.json)
- [`pascals-triangle`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/pascals-triangle/canonical-data.json)
- [`phone-number`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/phone-number/canonical-data.json)
- [`protein-translation`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/protein-translation/canonical-data.json)
- [`rotational-cipher`](https://github.com/exercism/problem-specifications/blob/4ab6ff38e4f3e96be2160360e30c606e94fdb5b6/exercises/rotational-cipher/canonical-data.json)

**Comment by ErikSchierboom** [comment]
I agree that the test case descriptions become very long, but I don't necessarily mind it _that_ much. The `tests.toml` files are rarely updated/edited manually anyway, so I don't think the verbosity hurts much. I don't really like the comment option as to me it does not help show the underlying structure better than the concatenated descriptions.

@SaschaMann @SleeplessByte what do you think?

**Comment by ee7** [comment]
Edited my post above to add more options.

But I think 3b is my favourite so far. I think with option 1:
- it's too hard to see the actual test name
- when two upper-level descriptions are the similar length, it's hard to see where one grouping ends and the next begins.

**Comment by SaschaMann** [comment]
I think I'd prefer 4, but with the descriptions of the test groups as table header names.

> I agree that the test case descriptions become very long, but I don't necessarily mind it that much. 

I agree. It also matches how nested tests are displayed on the website, at least by some test runners.

**Comment by ErikSchierboom** [comment]
I'm not a fan of the nested tables, as I find them to be hard to read.

@SleeplessByte @cmccandless Which option do you prefer?

**Comment by cmccandless** [comment]
The current implementation feels good enough to me. This file is generated and managed by configlet, right? If it's not meant to be edited manually, it doesn't matter if it's a little ugly to look at. Many machine-managed files are not optimized for human readability.

**Comment by ErikSchierboom** [comment]
> This file is generated and managed by configlet, right? 

That is correct. Maintainers would only manually update this file to either:

- Add an `include = false` property to indicate that a test isn't implemented (relatively uncommon)
- Add extra, custom properties (probably relatively uncommon)

**Comment by SleeplessByte** [comment]
I can't tell what the latest proposal is.

**Comment by ErikSchierboom** [comment]
@SleeplessByte It's option 1 in [this comment](https://github.com/exercism/configlet/pull/301#pullrequestreview-652127300). So basically, just appending the description path with ` → ` as the separator.

CC @angelikatyborska and @SaschaMann: I'd be interested in your thoughts too.

**Comment by angelikatyborska** [comment]
From options 1-4 suggested by ee7, I like option 1 (the initial implementation) the most because the parent description is right there on every test case (as opposed to option 3 where it only appears once) and there's no weird indentation.

> I'd also like to have a bikeshedding session on the separator character itself - I'm don't love →, but I don't know what's better.

Would it be possible to make `description` an array/list/whatever?


Instead a single log string:
```toml
[9a001b50-f194-4143-bc29-2af5ec1ef652]
description = "Check if the given string is an isogram → isogram with only lower case characters"
```

Split it like this? 
```toml
[9a001b50-f194-4143-bc29-2af5ec1ef652]
description = [
  "Check if the given string is an isogram",
  "isogram with only lower case characters"
]
```

**Comment by SleeplessByte** [comment]
I like proposal 1. I'm also like simplicity over nesting, and don't mind if the lines are long.

**Comment by ErikSchierboom** [comment]
Okay, then I propose to go with proposal 1 and use `→` as the separator.

@ee7 Could you review the code?

**Comment by ee7** [comment]
Sorry for the delay. There were a few PRs that touched the same code that I had to merge before getting to this one.

@ErikSchierboom I've fixed the tests and rebased this PR on `main`. One obstacle was that I'd refactored `ProbSpecsTestCase` to be a `distinct JsonNode` rather than an `object`, which meant I ended up changing the approach. Happy with the changes? Any thoughts?

@angelikatyborska Nice idea - I hadn't thought of that. A multi-line string would also be possible. But it seems like there's a consensus for the single line approach, so I'm happy to go with that.

To everyone, before I approve this: Are we happy with `→` as the delimiter? Should we use something like `->` instead?

`→` doesn't always render well (there might be clearer arrows), and it isn't easy for a user to type. Although I don't know why someone would want to type it - maybe when Ctrl+f or `git grep`ing for something.

**Comment by ErikSchierboom** [comment]
>  Happy with the changes?

Yes! Thanks.

> To everyone, before I approve this: Are we happy with → as the delimiter? Should we use something like -> instead?
> → doesn't always render well (there might be clearer arrows), and it isn't easy for a user to type. Although I don't know why someone would want to type it - maybe when Ctrl+f or git greping for something.

I don't really care much about the delimited. I'm fine with `->` too.

**Comment by SaschaMann** [comment]
I'd guess there's a higher risk of `->` ever showing up in descriptions given that unicode is somewhat frowned upon in prob-specs files due to tooling issues.

**Comment by ee7** [comment]
> I'd guess there's a higher risk of `->` ever showing up in descriptions given that unicode is somewhat frowned upon in prob-specs files due to tooling issues.

Yes. But maybe it's better to use `->` here, and then add a CI check to `problem-specifications` to forbid `->` inside `description`. What do you think?

If we use `→` in this repo then I guess I'd argue that we should forbid `→` in a prob-specs `description`, although it's smaller concern.

There aren't any `->` inside a `description` currently, and I don't know if a legimate use is likely to appear - perhaps a new exercise that parses some custom language that defines `->`?

**Comment by ErikSchierboom** [comment]
Maybe use `→` for now and update later if needed?

**Review by ee7 — APPROVED** [review]
Okay, I'll merge then. I'll make a follow-up PR later for the delimiter bikeshedding.

It's just that I want to avoid a mass PR that updates every `→` to e.g. `->` on every track. And I think that we should consider making a mass PR that adds `reimplements`, the fuller `description`, and the new comment at the top of every `tests.toml`. This would better explain how `tests.toml` works, and prevent noisy diffs in the first PR for every track exercise that touches the corresponding `tests.toml`.

So if we do want such a mass PR, I think we should settle the delimiter decision and the `tests.toml` comment wording first. We might also want to settle these things before creating the next configlet release.

**Comment by ErikSchierboom** [comment]
> So if we do want such a mass PR, I think we should settle the delimiter decision and the tests.toml comment wording first. We might also want to settle these things before creating the next configlet release.

Fine with me, as long as we keep the bikeshedding to a minimum. There is so much functionality left to implement that I'd rather we spend our time building things that bikeshedding over separators 🤷 But we can _probably_ decide upon a delimiter and description in relatively little time.

**Pull request merged**
